### PR TITLE
Transformations: Support escaped characters in key-value pair parsing

### LINKED ISSUE #47604: Extract Fields Transformer - Line splitting applying to values enclosed in quotes
As the Title says at the time of using Google Cloud Monitoring Datasource - Table Panel is treating words of one sentence as different Objects specially when there's a space, colon, comma are available within a sentence

Please have a look at this issue and assist with fixing it

- [Questions should be posted to: https://community.grafana.com](https://community.grafana.com/t/grafana-table-span/40885/6)
- How to record and attach gif: https://bit.ly/2Mi8T6K
Data Parsing

**Google Cloud Monitoring Datasource  - Table Panel Treating words of one sentence as different Objects**:

**Words should not be treated as separate Object but it instead treat whole sentence as one Object**:

**Datasource (Google Cloud Monitoring)>> Select a label which has a big sentence with multiple words and symbols from Stackdriver  metric >> Table Panel >> Transform >> Series to Row >> Extract to Fields (metric) **:

**Environment**:
- Grafana version: Grafana v8.3.1
- Data source type & version: Google Cloud Monitoring (Not sure of version, don't have access to view)
- OS Grafana is installed on: Linux
- User OS & Browser: Windows & Chrome
- Grafana plugins: Stackdriver/Google Cloud Monitoring
- Others: Use of Metrics, Table Panel > Transform

### PR DESCRIPTION
**Pull request by aangelisc**
**What this PR does / why we need it**:

This PR provides updated logic for the field extractor that allows for keys/values containing single or double quoted strings to be parsed appropriately when they include any potential character that may be used as a delimiter. 

**Which issue(s) this PR fixes**:

Fixes #47604 

**Special notes for your reviewer**:

Additional tests added for this specific use-case. Testing the correct parsing of quoted values and also values containing nested quotes.

Test was also carried out using the following steps:

1. Creating a custom metric in GCP with the following properties:
Metric Type: `counter`
Metric name: `extractor-test`
Filter: `resource.labels.project_id="PROJECT_NAME"`
Custom label: `Label name: test_fields,
Description: '',
Label type: STRING,
Field Name: textPayload`
2. The Cloud monitoring datasource was then configured to query `Logging` service of the specified project, selecting the metric name described above.
3.  Set the `Group by` value to the custom label name defined above.
4. Add the `Series to rows` transformation.
5. Add the `Extract fields` transformation, selecting the source as `Metric` and leaving the other parameters as default.
6. The `metric.label.test_fields` metric should be extracted as a single column with no extraneous fields extracted.

### CODE REVIEW

File: public/app/features/transformers/extractFields/fieldExtractor.test.ts [hunk 1]
```diff
@@ -14,6 +14,37 @@ describe('Extract fields from text', () => {
...
+        "x": "y",
+        "z": "d and 4",
+      }
+    `);
+  });
+
+  it('Test key-values with nested single/double quotes', async () => {
+    const extractor = fieldExtractors.get(FieldExtractorID.KeyValues);
+    const out = extractor.parse(
+      `a="1",   "b"=\'2\',c=3  x:y ;\r\nz="dbl_quotes="Double Quotes" sgl_quotes='Single Quotes'"`
```
Comment (ryantxu) [inline public/app/features/transformers/extractFields/fieldExtractor.test.ts:34]: this input seems wrong --- I would expect it to parse into:
`nz=dbl_quotes` 
and then giberish

**Comment by ryantxu** [comment]
I tried an alternate - non-regexp based solution.  See:
https://github.com/grafana/grafana/compare/key-value-reader#diff-601e183dc56c031bed480c34d0a002a06d6646439454811c379aaa153a3d5230R29

I'm not sure if it is better or just different :).  It does have the weird feature/bug that a string like "a b c d" becomes:
```
{ a="b", c="d" }
```

**Comment by ryantxu** [comment]
Do we really want to backport to 8.3?  for sure 8.5 -- I don't feel too strongly, but feels more like a feature improvement (support quotes and escaping) than a bug fix that should be aggressively back-ported

File: public/app/features/transformers/extractFields/fieldExtractor.test.ts [hunk 2]
```diff
@@ -14,6 +14,37 @@ describe('Extract fields from text', () => {
...
+    );
+
+    expect(out).toMatchInlineSnapshot(`
+      Object {
+        "a": "1",
+        "b": "2",
+        "c": "3",
+        "x": "y",
+        "z": "dbl_quotes=\\"Double Quotes\\" sgl_quotes=\\"Single Quotes\\"",
```
Comment (fridgepoet) [inline public/app/features/transformers/extractFields/fieldExtractor.test.ts:43]: `\\"Single Quotes\\""`

I think I'm seeing here that the single quotes are lost and replaced with double quotes.
What do you think about preserving the single quotes?

**Comment by fridgepoet** [comment]
From the original issue:
> treating words of one sentence as different Objects specially when there's a space, colon, comma

Could we also get some tests with quotes around a colon or comma and preserving those as one value?

File: public/app/features/transformers/extractFields/fieldExtractor.test.ts (hunk 1 above)
Comment (aangelisc) [reply public/app/features/transformers/extractFields/fieldExtractor.test.ts:34]: Have updated the test for this with the correct escaping

**Comment by aangelisc** [comment]
> I tried an alternate - non-regexp based solution. See: https://github.com/grafana/grafana/compare/key-value-reader#diff-601e183dc56c031bed480c34d0a002a06d6646439454811c379aaa153a3d5230R29
> 
> I'm not sure if it is better or just different :). It does have the weird feature/bug that a string like "a b c d" becomes:
> 
> ```
> { a="b", c="d" }
> ```

I think this solution may be better as it correctly handles separator characters that are enclosed in quotes and doesn't replace all quotes with double quotes by default when nested. Just looking into the bug you've mentioned and also another unusual case where if the input is something like `"a b='2',c=3  x:y"` the first character is dropped.


> From the original issue:
> 
> > treating words of one sentence as different Objects specially when there's a space, colon, comma
> 
> Could we also get some tests with quotes around a colon or comma and preserving those as one value?

Will add additional tests for these cases.

File: public/app/features/transformers/extractFields/fieldExtractor.test.ts (hunk 2 above)
Comment (aangelisc) [reply public/app/features/transformers/extractFields/fieldExtractor.test.ts:43]: We should be preserving single quotes, I could expand the regexp code to do this but I think @ryantxu's code may be more appropriate as it handles this case.

**Review by fridgepoet — APPROVED** [review]

**Comment by fridgepoet** [comment]
look good to me, will let @ryantxu approve
I guess the label needs to be changed to backport 8.5.x

**Review by ryantxu — APPROVED** [review]

**Pull request merged**